solr/core/src/java/org/apache/solr/handler/admin/api/GetSegmentData.java

```diff
@@ -0,0 +1,501 @@
+      boolean withFieldInfos,
+      List<LeafReaderContext> leafContexts,
+      IndexSchema schema)
+      throws IOException {
+    final var segmentInfo = new GetSegmentDataResponse.SingleSegmentData();
+    segmentInfo.name = segmentCommitInfo.info.name;
+    segmentInfo.delCount = segmentCommitInfo.getDelCount();
+    segmentInfo.softDelCount = segmentCommitInfo.getSoftDelCount();
+    segmentInfo.hasFieldUpdates = segmentCommitInfo.hasFieldUpdates();
+    segmentInfo.sizeInBytes = segmentCommitInfo.sizeInBytes();
+    segmentInfo.size = segmentCommitInfo.info.maxDoc();
+    Long timestamp = Long.parseLong(segmentCommitInfo.info.getDiagnostics().get("timestamp"));
+    segmentInfo.age = new Date(timestamp);
+    segmentInfo.source = segmentCommitInfo.info.getDiagnostics().get("source");
+    segmentInfo.version = segmentCommitInfo.info.getVersion().toString();
+
+    // don't open a new SegmentReader - try to find the right one from the leaf contexts
+    SegmentReader seg = null;
+    for (LeafReaderContext lrc : leafContexts) {
+      LeafReader leafReader = lrc.reader();
+      leafReader = FilterLeafReader.unwrap(leafReader);
+      if (leafReader instanceof SegmentReader sr) {
+        if (sr.getSegmentInfo().info.equals(segmentCommitInfo.info)) {
+          seg = sr;
+          break;
+        }
+      }
+    }
+    if (seg != null) {
+      LeafMetaData metaData = seg.getMetaData();
+      if (metaData != null) {
+        segmentInfo.createdVersionMajor = metaData.getCreatedVersionMajor();
+        segmentInfo.minVersion = metaData.getMinVersion().toString();
+        if (metaData.getSort() != null) {
+          segmentInfo.sort = metaData.getSort().toString();
+        }
+      }
+    }
+
+    if (!segmentCommitInfo.info.getDiagnostics().isEmpty()) {
+      segmentInfo.diagnostics =
+          SolrJacksonMapper.getObjectMapper()
+              .convertValue(
+                  segmentCommitInfo.info.getDiagnostics(),
+                  GetSegmentDataResponse.SegmentDiagnosticInfo.class);
+    }
+    if (!segmentCommitInfo.info.getAttributes().isEmpty()) {
+      segmentInfo.attributes = segmentCommitInfo.info.getAttributes();
+    }
+    if (withSizeInfo) {
+      Directory dir = segmentCommitInfo.info.dir;
+      List<Pair<String, Long>> files =
+          segmentCommitInfo.files().stream()
+              .map(
+                  f -> {
+                    long size = -1;
+                    try {
+                      size = dir.fileLength(f);
+                    } catch (IOException e) {
+                    }
+                    return new Pair<String, Long>(f, size);
+                  })
+              .sorted(
+                  (p1, p2) -> {
+                    if (p1.second() > p2.second()) {
+                      return -1;
+                    } else if (p1.second() < p2.second()) {
+                      return 1;
+                    } else {
+                      return 0;
+                    }
```

Comment:
I think this can be replaced by usage of `java.util.Comparator#comparing(java.util.function.Function<? super T,? extends U>)`

Reply:
[0] This stuff actually pre-exists this PR; it previously lived in SegmentInfoRequestHandler.

I generally try to avoid unrelated cleanups as I go through these v2 PRs. Partly for scope (if I stopped to fix every little thing, forward progress would halt entirely). And partly because you never know what seemingly "safe" change will end up biting you.